From: Yicong Yang <yangyicong@hisilicon.com>
Subject: watchdog/perf: provide function for adjusting the event period
Date: Thu, 19 Jun 2025 19:48:04 +0800

Patch series "Update the watchdog period according to real CPU frequency",
v2.

watchdog perf needs architecture to provide method for converting the
watchdog thresh to counter period.  For arm64 we're using the max CPU
frequency for doing the conversion which is from cpufreq driver.  But some
cpufreq driver are registered lately, for example cppc_cpufreq will be
registered at late initcall which is after the initialization of watchdog
perf (initialized in armv8_pmuv3 of device initcall).  In such case the
period of watchdog will not be accurate enough.  Fix this by registering a
cpufreq notifier and update the watchdog period once the cpufreq driver is
initialized.

Attach some Q&A raised by Andrew last time: [1]
https://lore.kernel.org/linux-arm-kernel/20250512160612.c10464075df3c7842b13da11@linux-foundation.org/

Q: What is the impact of this change?  Is the current code causing
   problems?  If so, what are they?  How is the end-user experience
   improved by this change?  Important info!

A: This will make NMI watchdog (hardlockup detector) work more
   accurately.  HARDLOCKUP_DETECTOR_PERF is driven by the PMU sample
   interrupts of cpu cycle event with period related to the watchdog
   threshold.  User will set the watchdog threshold in seconds (e.g.  10s
   by default) and for HARDLOCKUP_DETECTOR_PERF we need to convert the
   seconds to cycles to setup the PMU counter.  The coversion method is
   provided by the arhitecture by implementing hw_nmi_get_sample_period().
   For arm64 it's using max_cpufreq to do the conversion:

       cycle_event_period = threshold(s) * max_cpufreq(hz)

   Since arm64 doesn't have an architectural way to get the
   max_cpufreq, we use cpufreq_driver to get it.  If cpufreq_driver is not
   available, currently we use a safe max_cpufreq as 5GHz.  Without this
   patchset, if the cpufreq_driver is initialized after the hardlockup
   detector we'll use 5GHz for calculating the event period.  It's my case
   here as described in the coverletter.  That means in the default case
   (10s threshold) if the real max_cpufreq is 2.5GHz, the NMI watchdog is
   actually working in a 20s period.  With this patchset the period can be
   calibrated after the cpufreq driver is initialized, much more accurate.

Q: As far as I can tell, this patchset impacts arm64 only.  Do you
   think that other architectures should implement this?

A: It highly depends on the architecure's implementation of their
   HARDLOCKUP_DETECTOR_PERF (the implementation of
   hw_nmi_get_sample_period()).  If other architectures can gain the
   max_cpufreq without cpufreq driver they don't have this problems (seems
   x86 implement in another way and can gain the cpufreq by some
   arhitectural way).

Q: As far as I can tell, this patchset affects all cpufreq drivers
   which use late_initcall() (on arm64, of course).  Is this correct?

A: cpufreq drivers are not touched.  we registered a notifier block to
   the cpufreq framework so the changes to the cpufreq will notify us to
   see whether it's needed to modify the event period.

Q: It is asserted that we should use the *maximum* possible CPU
   frequency for this calculation.  Why?  I assume this is because we care
   about the minimum watchdog period?

A: Because it's impossbile to use the current frequency for calculating
   the counter period as it may change at any time.  Using maximum
   frequency will make the real period as close as possible to the
   expected watchdog thresh in a simplest way.


This patch (of 2):

Architectures using perf events for hard lockup detection need to convert
the watchdog_thresh to the event's period, some architectures (for example
arm64) perform this conversion using the CPU's maximum frequency which will
be acquired by cpufreq.

However when the lockup detector is initialized the cpufreq driver may not
be initialized, thus launching a watchdog with an inaccurate period.  

Provide a function hardlockup_detector_perf_adjust_period() to allowing
adjusting the event period.  Then the architecture can update with more
accurate period if cpufreq is initialized.

Link: https://lkml.kernel.org/r/20250619114805.16968-1-yangyicong@huawei.com
Link: https://lkml.kernel.org/r/20250619114805.16968-2-yangyicong@huawei.com
Signed-off-by: Yicong Yang <yangyicong@hisilicon.com>
Cc: "Rafael J. Wysocki" <rafael@kernel.org>
Cc: Viresh Kumar <viresh.kumar@linaro.org>
Cc: Catalin Marinas <catalin.marinas@arm.com>
Cc: Doug Anderson <dianders@chromium.org>
Cc: Jie Zhan <zhanjie9@hisilicon.com>
Cc: Joanthan Cameron <Jonathan.Cameron@huawei.com>
Cc: Lecopzer Chen <lecopzer.chen@mediatek.com>
Cc: Liu Pingfan <kernelfans@gmail.com>
Cc: Song Liu <song@kernel.org>
Cc: Sumit Garg <sumit.garg@linaro.org>
Cc: Thomas Gleinxer <tglx@linutronix.de>
Cc: Will Deacon <will@kernel.org>
Cc: Yicong Yang <yangyicong@hisilicon.com>
Cc: Zeng Tao <prime.zeng@hisilicon.com>

Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/nmi.h    |    2 ++
 kernel/watchdog_perf.c |   23 +++++++++++++++++++++++
 2 files changed, 25 insertions(+)

--- a/include/linux/nmi.h~watchdog-perf-provide-function-for-adjusting-the-event-period
+++ a/include/linux/nmi.h
@@ -103,10 +103,12 @@ void watchdog_hardlockup_check(unsigned
 extern void hardlockup_detector_perf_stop(void);
 extern void hardlockup_detector_perf_restart(void);
 extern void hardlockup_config_perf_event(const char *str);
+extern void hardlockup_detector_perf_adjust_period(int cpu, u64 period);
 #else
 static inline void hardlockup_detector_perf_stop(void) { }
 static inline void hardlockup_detector_perf_restart(void) { }
 static inline void hardlockup_config_perf_event(const char *str) { }
+static inline void hardlockup_detector_perf_adjust_period(int cpu, u64 period) { }
 #endif
 
 void watchdog_hardlockup_stop(void);
--- a/kernel/watchdog_perf.c~watchdog-perf-provide-function-for-adjusting-the-event-period
+++ a/kernel/watchdog_perf.c
@@ -187,6 +187,29 @@ void watchdog_hardlockup_disable(unsigne
 }
 
 /**
+ * hardlockup_detector_perf_adjust_period - Adjust the event period due
+ *                                          to cpu frequency change
+ * @cpu: The CPU whose event period will be adjusted
+ * @period: The target period to be set
+ */
+void hardlockup_detector_perf_adjust_period(int cpu, u64 period)
+{
+	struct perf_event *event = per_cpu(watchdog_ev, cpu);
+
+	if (!(watchdog_enabled & WATCHDOG_HARDLOCKUP_ENABLED))
+		return;
+
+	if (!event)
+		return;
+
+	if (event->attr.sample_period == period)
+		return;
+
+	if (perf_event_period(event, period))
+		pr_err("failed to change period to %llu\n", period);
+}
+
+/**
  * hardlockup_detector_perf_stop - Globally stop watchdog events
  *
  * Special interface for x86 to handle the perf HT bug.
_
